Controller tests

Natia-bit opened the pull request:
Test for controller level 

Note:
I havent changed the service layer for insert, as whatever I tried it still returned 200

Current method: 
```
    @Transactional
    public void insert(User user) {
        userDao.findByEmail(user.email()).
            ifPresentOrElse(
                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
                () -> userDao.insert(user));
    }
```

All three versions that I ve done still returned 200 on postman, even when I changed the return type on controller from void to `ResponceEntity<Void>` and `ResponceEntity<User>`

Simple try 1
```
    public ResponseEntity<Void> insert(User user) {
        Optional<User> temp = userDao.findByEmail(user.email());
        if (temp.isPresent()){
            throw new ResponseStatusException(HttpStatus.CONFLICT);
        }
        userDao.insert(user);
        return  ResponseEntity.status(HttpStatus.CREATED).build();
    }
```

Try 2
 ```
   public void insert(User user) {
        userDao.findByEmail(user.email()).
            ifPresentOrElse(
                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
                () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
    }
```

Try 3 using `.map`
```
    public void insert(User user) {
        userDao.findByEmail(user.email()).
            ifPresentOrElse(
                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
                () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
    }
```

I think its something simple and Im missing completely or I went to completly different direction

Commit:
restController tests setup and tests for getAll

Commit:
controller findById tests

Commit:
create new user test

Commit:
insert user tests

Commit:
delete tests

charles-van-wyk commented on src/test/java/cc/chipchop/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,264 @@
+            new User(3, "tree@test.com", "pass")));
+
+        mockMvc.perform(get("/api/users"))
+            .andDo(print())
```
You don't need the `print()` now that the code works it is mostly for debugging. The more tests you add to the service the more unusable logging becomes as it will just spam the output.

charles-van-wyk commented on src/test/java/cc/chipchop/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,264 @@
+            .andExpectAll(
+                status().isOk(),
+                content().contentType(MediaType.APPLICATION_JSON),
+                jsonPath("$.size()").value(3),
+                jsonPath("$[0].id").value(1),
+                jsonPath("$[0].email").value("one@test.com"),
+                jsonPath("$[0].password").value("password")
```
Nothing wrong with this approach, I am just listing this as an alternative so your information. You can also use the `.json` method and write the expected json as a string. Or you can create an instance of the response class and use the `objectMapper.writeValueAsString()` to change the object to a json string. All of these are viable solutions so just letting you know there are alternatives.

Natia-bit replied:
Got it.

charles-van-wyk commented on src/test/java/cc/chipchop/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,264 @@
+        User dude = new User(1, "dude@test.com", "java");
+        doThrow(new RuntimeException()).when(userService).insert(any(User.class));
+
+        ServletException ex = assertThrows(ServletException.class, () ->
+            mockMvc.perform(post("/api/users")
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .content(objectMapper.writeValueAsString(dude)))
+                .andReturn()
+        );
```
Hmm, this doesn't look right... Calling an endpoint should never throw an exception, it should return a proper response with a `500`

Natia-bit replied:
I know... this is the point I started hitting the wall. In the begining I tried 

```
        mockMvc.perform(post("/api/users")
            .contentType(MediaType.APPLICATION_JSON)
            .content(objectMapper.writeValueAsString(dude))
        ).andExpect(status().is5xxServerError());
```

as `.andExpect(status().is5xxServerError());` made more sense. But the error that I could not get passed over was `jakarta.servlet.ServletException: Request processing failed: java.lang.RuntimeException`. At this point the test failes.

charles-van-wyk replied:
Try throwing a different exception from the service like `SQLException` or one of spring's `DataAccessExceptions`

Natia-bit replied:
I tried both of them, and the ConnectionException too, and then the insert is being unhappy with the following error

```
org.mockito.exceptions.base.MockitoException: 
Checked exception is invalid for this method!
Invalid: java.net.ConnectException
```

Which when I read into it, from my undestanding it seems that this occures because Mockito doesn't allow throwing checked exceptions from a method that does not declare throwing them. The solution that I read was to add annotation` @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)` but I still dont undestand why, as when I hit the endpoint with postam when the database is down, it still throws 500

charles-van-wyk replied:
That annotation is to make a method as handling an exception and translating it to an appropriate status code. In your case you want to have a general "catch-all" handler. Sooooooo you were right 🤣 we can use `@ControllerAdvice` So what you can do is create a new class that you mark with `@ControllerAdvice` and add a method that will catch any unhandled exception. Explicitly not telling you HOW to do that

charles-van-wyk commented on src/test/java/cc/chipchop/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,264 @@
+    public void givenUpdateUser_whenDatabaseIsDown_thenReturn500() throws Exception {
+        doThrow(new RuntimeException()).when(userService).update(eq(1L), any(User.class));
+
+        ServletException ex = assertThrows(ServletException.class, () ->
+            mockMvc.perform(put("/api/users/{id}", 1)
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .content(objectMapper.writeValueAsString(new User(1, "freshdude@test.com", "java"))))
+                .andReturn()
+        );
```
Same here

charles-van-wyk commented on src/test/java/cc/chipchop/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,264 @@
+    public void givenDeleteUser_whenDatabaseIsDown_thenReturn500() throws Exception {
+        doThrow(new RuntimeException()).when(userService).delete(2);
+
+        ServletException ex = assertThrows(ServletException.class, () ->
+            mockMvc.perform(delete("/api/users/{id}", 2))
+                .andReturn()
+        );
```
Same here

charles-van-wyk commented:
> Test for controller level
> 
> Note: I havent changed the service layer for insert, as whatever I tried it still returned 200
> 
> Current method:
> 
> ```
>     @Transactional
>     public void insert(User user) {
>         userDao.findByEmail(user.email()).
>             ifPresentOrElse(
>                 u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
>                 () -> userDao.insert(user));
>     }
> ```
> 
> All three versions that I ve done still returned 200 on postman, even when I changed the return type on controller from void to `ResponceEntity<Void>` and `ResponceEntity<User>`
> 
> Simple try 1
> 
> ```
>     public ResponseEntity<Void> insert(User user) {
>         Optional<User> temp = userDao.findByEmail(user.email());
>         if (temp.isPresent()){
>             throw new ResponseStatusException(HttpStatus.CONFLICT);
>         }
>         userDao.insert(user);
>         return  ResponseEntity.status(HttpStatus.CREATED).build();
>     }
> ```
> 
> Try 2
> 
> ```
>   public void insert(User user) {
>        userDao.findByEmail(user.email()).
>            ifPresentOrElse(
>                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
>                () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
>    }
> ```
> 
> Try 3 using `.map`
> 
> ```
>     public void insert(User user) {
>         userDao.findByEmail(user.email()).
>             ifPresentOrElse(
>                 u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
>                 () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
>     }
> ```
> 
> I think its something simple and Im missing completely or I went to completly different direction

Apologies but can you give me a little more context here? 🙈 When you say "All three still return 200" what do you mean? The best way to describe or give people context when describing problems is like naming a test case:
- Given something I'm doing
- When I do `something` I expect `this`
- But what is actually happening is `that`

So to make it more concrete with your examples here:
- I'm refactoring the insert method to one of the other suggestions we talked about
- I'm hitting the endpoint via postman twice with the same data and that should give me a 200 first and a 409 on the second call
- It is always giving me a 200 no matter how many times I do it or how I refactor the code

Natia-bit commented:
> > Test for controller level
> > Note: I havent changed the service layer for insert, as whatever I tried it still returned 200
> > Current method:
> > ```
> >     @Transactional
> >     public void insert(User user) {
> >         userDao.findByEmail(user.email()).
> >             ifPresentOrElse(
> >                 u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
> >                 () -> userDao.insert(user));
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > All three versions that I ve done still returned 200 on postman, even when I changed the return type on controller from void to `ResponceEntity<Void>` and `ResponceEntity<User>`
> > Simple try 1
> > ```
> >     public ResponseEntity<Void> insert(User user) {
> >         Optional<User> temp = userDao.findByEmail(user.email());
> >         if (temp.isPresent()){
> >             throw new ResponseStatusException(HttpStatus.CONFLICT);
> >         }
> >         userDao.insert(user);
> >         return  ResponseEntity.status(HttpStatus.CREATED).build();
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Try 2
> > ```
> >   public void insert(User user) {
> >        userDao.findByEmail(user.email()).
> >            ifPresentOrElse(
> >                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
> >                () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
> >    }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Try 3 using `.map`
> > ```
> >     public void insert(User user) {
> >         userDao.findByEmail(user.email()).
> >             ifPresentOrElse(
> >                 u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
> >                 () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think its something simple and Im missing completely or I went to completly different direction
> 
> Apologies but can you give me a little more context here? 🙈 When you say "All three still return 200" what do you mean? The best way to describe or give people context when describing problems is like naming a test case:
> 
> * Given something I'm doing
> * When I do `something` I expect `this`
> * But what is actually happening is `that`
> 
> So to make it more concrete with your examples here:
> 
> * I'm refactoring the insert method to one of the other suggestions we talked about
> * I'm hitting the endpoint via postman twice with the same data and that should give me a 200 first and a 409 on the second call
> * It is always giving me a 200 no matter how many times I do it or how I refactor the code

Ok, sorry about that.  Lets try again 
When a user is successfuly created using the `insert()` method from `UserService`, the current repsponce code is 200. 
My goal is to change the responce code from 200 to 201. 

I tried the following 3 ways to return 201, and in all the options, I used `HttpStatus.CREATED` 
I changed the return type in controller to return `ResponseEntity<Void>`
When testing each of the three methods with Postman, each time a user was successfully created, it returned 200 and not 201

Controller.java
```
    @PostMapping("/users")
    public ResponseEntity<Void> createUser(@RequestBody User user) {
        userService.insert(user);
    }

```

UserService.java 
```
    @Transactional
    public void insert(User user) {
        userDao.findByEmail(user.email()).
            ifPresentOrElse(
                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
                () -> userDao.insert(user));
    }
```

UserService.java  - Simple try 1
```
    public ResponseEntity<Void> insert(User user) {
        Optional<User> temp = userDao.findByEmail(user.email());
        if (temp.isPresent()){
            throw new ResponseStatusException(HttpStatus.CONFLICT);
        }
        userDao.insert(user);
        return  ResponseEntity.status(HttpStatus.CREATED).build();
    }
```

UserService.java  - Try 2
```
  public void insert(User user) {
       userDao.findByEmail(user.email()).
           ifPresentOrElse(
               u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
               () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
   }
```

UserService.java - Try 3 using .map
```
    public void insert(User user) {
        userDao.findByEmail(user.email()).
            ifPresentOrElse(
                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
                () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
    }
```

I think all three are minor variations of each other, and I get the feeling Im missing something

charles-van-wyk commented:
> > > Test for controller level
> > > Note: I havent changed the service layer for insert, as whatever I tried it still returned 200
> > > Current method:
> > > ```
> > >     @Transactional
> > >     public void insert(User user) {
> > >         userDao.findByEmail(user.email()).
> > >             ifPresentOrElse(
> > >                 u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
> > >                 () -> userDao.insert(user));
> > >     }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > All three versions that I ve done still returned 200 on postman, even when I changed the return type on controller from void to `ResponceEntity<Void>` and `ResponceEntity<User>`
> > > Simple try 1
> > > ```
> > >     public ResponseEntity<Void> insert(User user) {
> > >         Optional<User> temp = userDao.findByEmail(user.email());
> > >         if (temp.isPresent()){
> > >             throw new ResponseStatusException(HttpStatus.CONFLICT);
> > >         }
> > >         userDao.insert(user);
> > >         return  ResponseEntity.status(HttpStatus.CREATED).build();
> > >     }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Try 2
> > > ```
> > >   public void insert(User user) {
> > >        userDao.findByEmail(user.email()).
> > >            ifPresentOrElse(
> > >                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
> > >                () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
> > >    }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Try 3 using `.map`
> > > ```
> > >     public void insert(User user) {
> > >         userDao.findByEmail(user.email()).
> > >             ifPresentOrElse(
> > >                 u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
> > >                 () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
> > >     }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I think its something simple and Im missing completely or I went to completly different direction
> > 
> > 
> > Apologies but can you give me a little more context here? 🙈 When you say "All three still return 200" what do you mean? The best way to describe or give people context when describing problems is like naming a test case:
> > 
> > * Given something I'm doing
> > * When I do `something` I expect `this`
> > * But what is actually happening is `that`
> > 
> > So to make it more concrete with your examples here:
> > 
> > * I'm refactoring the insert method to one of the other suggestions we talked about
> > * I'm hitting the endpoint via postman twice with the same data and that should give me a 200 first and a 409 on the second call
> > * It is always giving me a 200 no matter how many times I do it or how I refactor the code
> 
> Ok, sorry about that. Lets try again When a user is successfuly created using the `insert()` method from `UserService`, the current repsponce code is 200. My goal is to change the responce code from 200 to 201.
> 
> I tried the following 3 ways to return 201, and in all the options, I used `HttpStatus.CREATED` I changed the return type in controller to return `ResponseEntity<Void>` When testing each of the three methods with Postman, each time a user was successfully created, it returned 200 and not 201
> 
> Controller.java
> 
> ```
>     @PostMapping("/users")
>     public ResponseEntity<Void> createUser(@RequestBody User user) {
>         userService.insert(user);
>     }
> ```
> 
> UserService.java
> 
> ```
>     @Transactional
>     public void insert(User user) {
>         userDao.findByEmail(user.email()).
>             ifPresentOrElse(
>                 u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
>                 () -> userDao.insert(user));
>     }
> ```
> 
> UserService.java - Simple try 1
> 
> ```
>     public ResponseEntity<Void> insert(User user) {
>         Optional<User> temp = userDao.findByEmail(user.email());
>         if (temp.isPresent()){
>             throw new ResponseStatusException(HttpStatus.CONFLICT);
>         }
>         userDao.insert(user);
>         return  ResponseEntity.status(HttpStatus.CREATED).build();
>     }
> ```
> 
> UserService.java - Try 2
> 
> ```
>   public void insert(User user) {
>        userDao.findByEmail(user.email()).
>            ifPresentOrElse(
>                u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
>                () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
>    }
> ```
> 
> UserService.java - Try 3 using .map
> 
> ```
>     public void insert(User user) {
>         userDao.findByEmail(user.email()).
>             ifPresentOrElse(
>                 u -> {throw new ResponseStatusException(HttpStatus.CONFLICT);},
>                 () -> ResponseEntity.status(HttpStatus.CREATED).body(userDao.insert(user)));
>     }
> ```
> 
> I think all three are minor variations of each other, and I get the feeling Im missing something

Okay cool, I got you! So you need to indicate that on the controller not the service. Since you always want to the return 201 you can annotate the controller method with `@ResponseStatus(HttpStatus.CREATED)` or return `ResponseEntity.status(HttpStatus.CREATED)` from the controller. Both should work

Commit:
changing tests - before controller advise

Commit:
tests with ControllerExpectionHandler

Commit:
when creating user controller returns 201. Controller test modified too

Natia-bit commented on src/main/java/cc/chipchop/rest/ChipchopRestController.java:
```diff
@@ -26,8 +27,8 @@ public Optional<User> findUserById(@PathVariable long id) {
 
-
     @PostMapping("/users")
+    @ResponseStatus(HttpStatus.CREATED)
```
Was it this simple?

charles-van-wyk replied:
Yep, it's easy if you have a single status code to return. If you want to return `201` when it created a record and `200` when it's already existing then it becomes slightly more involved.

Natia-bit commented on src/test/java/cc/chipchop/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,274 @@
+    }
+
+    @Test
+    public void givenGetAllUsers_whenDatabaseIsDown_thenReturn503() throws Exception {
```
It made sense in my head but Im not sure of 503 is the most suitable to use?

charles-van-wyk replied:
You can return just 500 which is something internal broke 🤣

charles-van-wyk commented on src/main/java/cc/chipchop/exception/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,21 @@
+
+    @ExceptionHandler({DataAccessException.class, SQLException.class, ConnectException.class})
+    public ResponseEntity<String> handleDatabaseDown(Exception e){
+        return new ResponseEntity<>( HttpStatus.SERVICE_UNAVAILABLE );
```
Not sure how this looks in the api response at the end, but just make sure it follows the same structure as other exceptions.

charles-van-wyk commented on src/test/java/cc/chipchop/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,274 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@Import(ControllerExceptionHandler.class)
```
Don't know `@Import` I usually use `@ContextConfiguration` which is from the test package. But I can't tell you wha the difference is

charles-van-wyk commented on src/test/java/cc/chipchop/controller/UserRestControllerTest.java:
```diff
@@ -0,0 +1,274 @@
+    public void givenGetAllUsers_whenDatabaseIsDown_thenReturn500() throws Exception {
+        when(userService.findAll()).thenThrow(new RuntimeException());
+
+        assertThrows(ServletException.class, () ->
```
This test should hopefully fail now? We don't want any errors to be thrown from the controller. It should ALWAYS return a json response.

If it doesn't you might want to add a "catch-all" method to your controller advice that handles like `Exception.class` which is the base class for all exceptions in java

charles-van-wyk reviewed (commented):
Looks good, just some tiny comments. Almost there

Commit:
pr review changes
